Add additional version info to os.uname().version.

dpgeorge opened the pull request:
This additional info can be used to tell the exact git commit from this repository that was used to build the firmware.

Exampe output:
```
MicroPython v1.7-9-gbe020eb on 2016-04-18; micro:bit with nRF51822
Type "help()" for more information.
>>> import os
>>> os.uname().version
"micro:bit v1.0-b'64a61db' (dirty); MicroPython v1.7-9-gbe020eb on 2016-04-18"
>>> 
```

This gives the git hash of this repo, as well as the upstream uPy version.

If one wanted to extract this info from a hex file it would be pretty straightforward: just write a script to scan the .hex file for the string "micro:bit v" and then extract the version info.

See #370 for discussion around this point.

whaleygeek commented:
'microbit v' as a string sounds a little brittle.

For PXT we have a binary signature that identifies a metadata record, and that is pretty robust.

I appreciate this comment here is a retro-fit, but I wonder if longer term we need a more robust method of embedding simple metadata into the .hex so that it can be retrieved by an external tool?

I'm reminded of similar systems I developed years ago, where a specific binary tag and record termination protocol was used to allow any module to embed metadata into the final object - and I wrote a tool that scanned and extracted all tags, which was great for post-analysis of 'any old .hex file that happened to be lying around'. Allowing any module to use this tagging format was very powerful. The number of bytes required for a tag can be quite small, providing there are additional protections around the record, such as a defined protocol for termination of the record, it makes it quite robust.

dpgeorge commented:
> 'microbit v' as a string sounds a little brittle.

You can scan for more than that, you can also check for the "; MicroPython v" string and then verify that the whole thing has the right format (eg digits in the correct place, etc).

> I appreciate this comment here is a retro-fit

Yes this is a short term solution.  When (if?) we move to the mainline MicroPython repository then things will be different.

markshannon commented:
This should be very useful pulling apart issues like #392 and #421 in the future.
Happy to merge unless @whaleygeek has any objections.

whaleygeek commented:
No objections - anything that improves traceability to the precise source
files used to generate that .hex is always a good thing, thanks.

It will significantly improve our ability to send you much better bug
reports in the future (and filter out spurious ones that just need people
to upgrade to the latest deployed editors, or wait for planned fixes to be
deployed into those editors).


On 26 March 2017 at 11:35, Mark Shannon <notifications@github.com> wrote:

> This should be very useful pulling apart issues like #392
> <https://github.com/bbcmicrobit/micropython/issues/392> and #421
> <https://github.com/bbcmicrobit/micropython/issues/421> in the future.
> Happy to merge unless @whaleygeek <https://github.com/whaleygeek> has any
> objections.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/bbcmicrobit/micropython/pull/420#issuecomment-289271546>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AD5cASKU01tax-mlNaV8siJmIBq7Tsptks5rpj-PgaJpZM4MmMLH>
> .
>

markshannon commented:
I tried this out, but it makes no difference for me. `YOTTA_BUILD_VCS_DESCRIPTION` is not defined.

My `YOTTA_BUILD_INFO_HEADER` file looks this:
```c
#ifndef __YOTTA_BUILD_INFO_H__
#define __YOTTA_BUILD_INFO_H__
// yotta build info, #include YOTTA_BUILD_INFO_HEADER to access
#define YOTTA_BUILD_YEAR 2017 // UTC year
#define YOTTA_BUILD_MONTH 4 // UTC month 1-12
#define YOTTA_BUILD_DAY 9 // UTC day 1-31
#define YOTTA_BUILD_HOUR 16 // UTC hour 0-24
#define YOTTA_BUILD_MINUTE 26 // UTC minute 0-59
#define YOTTA_BUILD_SECOND 57 // UTC second 0-61
#define YOTTA_BUILD_UUID 9ffbacea-d18a-45dc-a0ca-0418b53ce567 // unique random UUID for each build
#define YOTTA_BUILD_VCS_ID e90c2dbc9273a55ad4053404565b2e27fd4e82fc // git or mercurial hash
#define YOTTA_BUILD_VCS_CLEAN 1 // evaluates true if the version control system was clean, otherwise false
#endif // ndef __YOTTA_BUILD_INFO_H__
```
Should we be using `YOTTA_BUILD_VCS_ID` instead?

dpgeorge commented:
That's strange, I have the same data in build/bbc-microbit-classic-gcc-nosd/yotta_build_info.h plus one extra line:
```
#define YOTTA_BUILD_VCS_DESCRIPTION b'9aa6dc9' // git describe or mercurial equivalent
```
Maybe I have a newer version of yotta?  yt --version gives 0.16.4 for me.  Also, my version has the VCS_ID as a Python bytes object:
```
#define YOTTA_BUILD_VCS_ID b'9aa6dc9ebf12700b9f282ac9d6ef6fd8a16ad26e' // git or mercurial hash
```

markshannon commented:
I had an ancient version of yotta, so I've upgraded (to 0.18.2) and it now works.
I think we should remove the `#if defined(YOTTA_BUILD_VCS_DESCRIPTION)` line, that way we can't accidentally produce a build without the git-tag (as was happening for me before I upgraded).

dpgeorge commented:
> I think we should remove the #if defined(YOTTA_BUILD_VCS_DESCRIPTION) line,

Ok, I've removed it (the downside is that you must have used git to check out the code).

I also added the build date to the version info.

deshipu commented:
You are potentially going to have a hard time with distro packagers with this
if you ever want it to get packaged into any distribution. The packages
are always built from released archives, and never from a git repo.

dpgeorge commented:
> The packages are always built from released archives, and never from a git repo.

We don't have release archives, and using yotta makes it hard to do that (unless there's a special yotta command to make a release archive...).  So, as long as we are tied to yotta (which hopefully won't be for much longer) I don't think it's a problem to use this build info.

markshannon commented:
LGTM

@dpgeorge, @deshipu are you both happy for me to merge this?

deshipu commented:
I see no further problems other than the one I mentioned, and as @dpgeorge said, that is a problem that we have anyways, at least for now. we can always change this if we decide to change the current build system.

dpgeorge commented:
Should be ok.
